### PR TITLE
Extend Permissions for viewing the dashboard.

### PR DESCRIPTION
**Pull request by kuasha420**
## Summary

Addresses issue:

- #4599

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: tests/phpunit/integration/Core/Permissions/PermissionsTest.php [hunk 1]
```diff
@@ -295,11 +297,14 @@ private function verify_view_shared_dashboard_capability( $author, $contributor
 		// item dismissed to VIEW_SHARED_DASHBOARD.
 		$this->assertFalse( user_can( $author, Permissions::VIEW_SHARED_DASHBOARD ) );
 		$this->assertFalse( user_can( $contributor, Permissions::VIEW_SHARED_DASHBOARD ) );
+		$this->assertFalse( user_can( $contributor, Permissions::VIEW_DASHBOARD ) );
 
 		$contributor_user_options = new User_Options( $this->context, $contributor->ID );
 		$dismissed_items          = new Dismissed_Items( $contributor_user_options );
 		$dismissed_items->add( 'shared_dashboard_splash', 0 );
 		$this->assertTrue( user_can( $contributor, Permissions::VIEW_SHARED_DASHBOARD ) );
+		// User should also be able to access VIEW_DASHBOARD as he has the VIEW_SHARED_DASHBOARD access.
```
Comment (techanvil) [inline tests/phpunit/integration/Core/Permissions/PermissionsTest.php:306]: I think this comment should be tweaked slightly to be in line with our inclusive language guidelines
```suggestion
		// User should also be able to access VIEW_DASHBOARD as they have the VIEW_SHARED_DASHBOARD access.
```

**Review by techanvil — CHANGES_REQUESTED** [review]
Hey @kuasha420, this one took me a little while to get my head around the permissions, but I've gone through the logic and the tests and it's LGTM.

One general comment, as you've had to go a bit beyond the IB, it would have been nice to write a note about that, maybe under **Relevant technical choices** section in the PR description. No biggie, just something to consider if you have to make some on the hoof changes next time.

I've left some super trivial comments to address but apart from that, this looks good to go. Cheers!

**Comment by kuasha420** [comment]
@techanvil Thanks for taking the time to review this one, yeah definitely could've done a better job at the IB as well as an explanation. I think most of the improvised stuffs were based on and built upon work done by @jimmymadon on previous tickets and those were in Execution while I was doing the IB for this one so some detailed were missing. 

Regardless, I have fixed up the typos. 

On something more substantial that both of us missed, There's a point in the AC (1st subpoint of 1st point) that calls for the `VIEW_POST_INSIGHT` to follow the same logic as the new `VIEW_DASHBOARD` logic. I have made a couple of change to achieve that but I'm not sure if I'm doing the right thing or if both changes are needed at all. Can you have a look at it please? 

Cheers.

**Comment by techanvil** [comment]
> On something more substantial that both of us missed, There's a point in the AC (1st subpoint of 1st point) that calls for the `VIEW_POST_INSIGHT` to follow the same logic as the new `VIEW_DASHBOARD` logic. I have made a couple of change to achieve that but I'm not sure if I'm doing the right thing or if both changes are needed at all. Can you have a look at it please?

Hi @kuasha420, thanks for pointing this out. I've been trying to get a grasp of this but I am having a bit of trouble figuring this one out myself.

Looking at the codebase _prior_ to your change, I can't quite figure `VIEW_POST_INSIGHTS` out - it seems to map to `edit_post` and deny access if the user doesn't have permission to access the post, but it also has a mapping to `VIEW_POSTS_INSIGHTS`, which looks as though it should then allow access if the user can `edit_posts`. However when I try to exercise this in a test, even though `user_can VIEW_POSTS_INSIGHTS`, a call to `user_can VIEW_POST_INSIGHTS` fails:

```php
$post_id = $this->factory()->post->create();
$this->assertTrue( user_can( $contributor, Permissions::VIEW_POSTS_INSIGHTS ) ); // This passes
$this->assertTrue( user_can( $contributor, Permissions::VIEW_POST_INSIGHTS, $post_id ) ); // But this fails
```

Therefore, reading the point in the AC, I am unable to fully understand the requirement. Should `VIEW_POST_INSIGHTS` allow access when the user can edit the specific post, _or_ when the user can view the main/authenticated dashboard or the shared dashboard - and does the present behaviour of it mapping to `VIEW_POSTS_INSIGHTS` still have relevance (and how is it supposed to actually work)?

Sorry if I have just added to the confusion here!

@aaemnnosttv, as you have experience of these permissions and it looks like you worked on the AC here, would you be able to shed a bit of light on this one, and possibly provide some pointers as to the implementation too?

cc @felixarntz @eugene-manuilov

**Comment by aaemnnosttv** [comment]
> Looking at the codebase _prior_ to your change, I can't quite figure `VIEW_POST_INSIGHTS` out - it seems to map to `edit_post` and deny access if the user doesn't have permission to access the post, but it also has a mapping to `VIEW_POSTS_INSIGHTS`, which looks as though it should then allow access if the user can `edit_posts`. 

@techanvil that's correct. Normally capabilities are either primitive (e.g. `edit_posts`) or meta (e.g. `edit_post` with ID). `VIEW_POSTS_INSIGHTS` seems to be somewhat of a special case in that it's used as both (as primitive in JS but meta in PHP – see `Admin_Bar::is_active`).

> However when I try to exercise this in a test, even though `user_can VIEW_POSTS_INSIGHTS`, a call to `user_can VIEW_POST_INSIGHTS` fails:
> 
> ```php
> $post_id = $this->factory()->post->create();
> $this->assertTrue( user_can( $contributor, Permissions::VIEW_POSTS_INSIGHTS ) ); // This passes
> $this->assertTrue( user_can( $contributor, Permissions::VIEW_POST_INSIGHTS, $post_id ) ); // But this fails
> ```
> 
> Therefore, reading the point in the AC, I am unable to fully understand the requirement. Should `VIEW_POST_INSIGHTS` allow access when the user can edit the specific post, _or_ when the user can view the main/authenticated dashboard or the shared dashboard - and does the present behaviour of it mapping to `VIEW_POSTS_INSIGHTS` still have relevance (and how is it supposed to actually work)?

This is where things get a little confusing, indeed. Here the meta cap check will resolve to `edit_post` with the given ID as you might expect. What you probably wouldn't expect (I learned this too) is that `edit_post` is itself a bit of a proxy for more specific capabilities based on the post status and the author of the post. Creating a post with the default factory arguments will create a **published** post with the current user as the author of the post (although this can be set explicitly by passing `post_author => $user_id` in the args when calling `create`. You can see how this breaks down in the dark depths of the `map_meta_cap` function: https://github.com/WordPress/WordPress/blob/b2253ef74cc5ab8a813269c21df64a2ca5434487/wp-includes/capabilities.php#L180-L205 

If your `$contributor` user in a user with the Contributor role, they won't be able to `edit_post` even if they are the author of the post, because contributors can't edit published posts – but an Author can. Similarly, an Author cannot edit another user's post, but an Editor can. When testing with a published post, you should see that both authors and editors should be able to `VIEW_POSTS_INSIGHTS`. 

See https://wordpress.org/support/article/roles-and-capabilities/#capability-vs-role-table for more info.

---

Regarding the PR, there are a few things to point out so I'll leave a review as well shortly but hopefully this sheds some more light on things.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @kuasha420 ! This looks pretty close – see my comments below.

**Comment by felixarntz** [comment]
@techanvil @aaemnnosttv Trying to clarify a bit related to https://github.com/google/site-kit-wp/pull/5031#issuecomment-1097182028:
* Note there is `VIEW_POSTS_INSIGHTS` and `VIEW_POST_INSIGHTS`.
* The first is a primitive capability that will rely on the core primitive capability `edit_posts`.
* The latter is a meta capability that will rely on our own primitive capability `VIEW_POSTS_INSIGHTS` as well as on the core meta capability `edit_post`.
* In other words, `current_user_can( VIEW_POST_INSIGHTS, 3 )` should be true if `current_user_can( VIEW_POSTS_INSIGHTS ) && current_user_can( 'edit_post', 3 )`.

Hopefully that helps.

**Comment by kuasha420** [comment]
@aaemnnosttv Thanks for the feedback. I have changed the `VIEW_SPLASH` capability to be a general one instead of `dashboardSharing` exclusive and simplified the logic in the `Screens`.

Also thank you and @felixarntz for trying to explain the `VIEW_POST_INSIGHTS`. I think I understand it a bit better. I'm still confused about the actions required to satisfy this though. Should I map the `VIEW_POSTS_INSIGHTS` to `VIEW_DASHBOARD` when `dashboardSharing` is enabled or `self::VIEW_POST_INSIGHTS            => self::VIEW_DASHBOARD,` is sufficient? 

Cheers.

**Comment by techanvil** [comment]
Thanks, @aaemnnosttv & @felixarntz, for explaining `VIEW_POST_INSIGHTS`, I definitely understand the workings of it better now. However I was about to write along the same lines as @kuasha420 has - I'm still a bit unclear on how it should be modified to satisfy the AC point. I will watch with interest for the answer to @kuasha420's question! :)

**Comment by aaemnnosttv** [comment]
> I guess I am not quite clear on the AC point itself and the intent behind it. From what I can see, `VIEW_POST_INSIGHTS` is only used when determining whether to show the SK Admin Bar, for a given post. At present, only users who can edit the post will see the Admin Bar. Do we want to expand this to users who can edit the post OR users who can view the authenticated/shared dashboard? That is the only way I can make sense of it, as it doesn't seem to add anything to further restrict `VIEW_POST_INSIGHTS` by adding an AND to the condition. However if we want to make it an OR then presumably we need to rewrite the `VIEW_POST_INSIGHTS` code a bit, making the `edit_post` mapping conditional...

@techanvil I'll attempt to clarify and @felixarntz can correct me if I'm wrong 😄 

Firstly, as Felix pointed out, there are two very similar post-related capabilities:
- `VIEW_POSTS_INSIGHTS` (primitive)
- `VIEW_POST_INSIGHTS` (meta – post-specific)

So `VIEW_POSTS_INSIGHTS` which is more generic is essentially checking if users can access SK data for posts (in general). That means they should be able to `VIEW_DASHBOARD` (the highest-level view of data about the site overall) but can also `edit_posts`. 

`VIEW_POST_INSIGHTS` (the post-specific capability) would require the generic `VIEW_POSTS_INSIGHTS` but also that the user can edit the specific post given to see its SK insights/data. So for lower capability users such as a contributor, they can edit posts, but they wouldn't be able to see SK insights for published posts (even their own), but an Author could. An Author could see insights for their own published posts, but not for other's posts. An editor could see insights for all posts (as could admins of course).

Hopefully that helps, let me know if you still have questions.

**Comment by felixarntz** [comment]
@kuasha420 I discussed this with @aaemnnosttv and @techanvil for a bit more. First of all, we should clarify that the `VIEW_POST_INSIGHTS` in the ACs was a typo - it should have said `VIEW_POSTS_INSIGHTS`, this has been corrected now. In other words, this PR should not make any changes to `VIEW_POST_INSIGHTS` at all.

Now to the heavy stuff. The way WordPress permissions work isn't exactly trivial, and some parts of the current class may make that harder to understand. To outline the general flow:
* Whenever you call `user_can()` or `current_user_can()`, it first goes into the `map_meta_cap` filter callback we have.
* In that filter callback, we can run whatever logic we need to map the one requested capability to a list/array of "base capabilities". If we want the permission check to fail without any actual capability checks, we can include `do_not_allow` in the list, which means the check will for sure result in `false`.
* After `map_meta_cap`, we now have a list/array of base capabilities, potentially including `do_not_allow`. If it includes `do_not_allow`, WordPress will just return `false` for the overall check. Otherwise, WordPress will check if the user has **all** of the capabilities in the list and only in that case return `true`.
    * The base capabilities of WordPress core are stored in the database per user role. However, it's a best practice for plugins to not store their own capabilities there too as there are some problems with that. That's where the `user_has_cap` filter comes in. In that filter, we can then map any of our custom base capabilities to core base capabilities from the database (that's what the `$base_to_core` class variable is for).

One further general note: Because all capabilities returned will be checked for whether the user has them, it should not be necessary for us to ever call `user_can()` from within our `map_meta_cap` filter callback. For example, instead of calling `user_can( $user_id, self::AUTHENTICATE )`, we should rather include `self::AUTHENTICATE` in the list of capabilities to check for afterwards. (There was a `user_can()` call in this class prior to this PR, but even that could be reworked - doesn't need to happen here though.)

For the next steps of this PR specifically, we came up with the following:
* Revert any changes related to `VIEW_POST_INSIGHTS`.
* Remove both `VIEW_AUTHENTICATED_DASHBOARD` and `VIEW_SHARED_DASHBOARD` from the `$base_to_core` list, as `VIEW_DASHBOARD` is sufficient as a base capability. That is because:
    * Checking for `VIEW_AUTHENTICATED_DASHBOARD` is the same as checking `VIEW_DASHBOARD` and `AUTHENTICATE`.
    * Checking `VIEW_SHARED_DASHBOARD` is the same as checking `VIEW_DASHBOARD`.
    * Of course there are still differences between them, but those only become relevant for logic in `map_meta_cap`.
* A similar thing applies to `VIEW_SPLASH`, we can remove that from `$base_to_core` as well.
* Detail: We also need to remove `VIEW_AUTHENTICATED_DASHBOARD` and `VIEW_SPLASH` from `$network_to_base` for the same reason.
* Per the above, since `VIEW_SHARED_DASHBOARD` is now no longer a base capabilitiy, we need to adjust the `READ_SHARED_MODULE_DATA` mapping in `$meta_to_base` to map to `VIEW_DASHBOARD` instead.
* Also per the above, `VIEW_AUTHENTICATED_DASHBOARD`, `VIEW_SHARED_DASHBOARD`, and `VIEW_SPLASH` can now be handled via `$meta_to_base`, so they should be added to that map.
    * `VIEW_AUTHENTICATED_DASHBOARD` should map to `array( VIEW_DASHBOARD, AUTHENTICATE )`.
    * `VIEW_SHARED_DASHBOARD` should map to `VIEW_DASHBOARD`.
    * `VIEW_SPLASH` should map to `VIEW_DASHBOARD`.
* Now let's get to the `map_meta_cap` logic:
    * There's already a `check_view_shared_dashboard_capability( $user_id )` method that checks for all the non-capability-related conditions around accessing the shared dashboard. We should introduce a corresponding `check_view_authenticated_dashboard_capability( $user_id )` method that simply checks if the user is authenticated and return `array( AUTHENTICATE )` if yes or `array( 'do_not_allow' )` if not.
    * Your `check_view_dashboard_capability( $user_id )` method should be modified as follows:
        * Call `check_view_authenticated_dashboard_capability( $user_id )`. If that return value includes `do_not_allow`, discard that and call `check_view_shared_dashboard_capability( $user_id )` instead. This ensures that at least one of them must not include `do_not_allow` for the overall check to be successful.
    * Your `check_view_splash_capability( $user_id )` needs some adjustments, most importantly it shouldn't rely on `user_can`:
        * If `dashboardSharing` feature flag is enabled: Call `check_view_shared_dashboard_capability( $user_id )`. If that includes `do_not_allow`, simply return `array( AUTHENTICATE )` since that will result in the `AUTHENTICATE` capability being required from the user.
        * If `dashboardSharing` feature flag is not enabled: Simply return `array( AUTHENTICATE )`, without any conditions.
    * In our `map_meta_cap` method we should then call `check_view_authenticated_dashboard( $user_id )` in case of $cap being `VIEW_AUTHENTICATED_DASHBOARD` and merge its return into `$caps`.
    * In case of `$cap` being one of `VIEW_DASHBOARD` or `VIEW_POSTS_INSIGHTS` we need to call the above `check_view_dashboard_capability( $user_id )`, as both of these caps should basically rely on the same.
    * In case of `$cap` being `VIEW_SPLASH` we can keep the condition you currently have, based on the adjustments to `check_view_splash_capability( $user_id )` mentioned above.

I understand this is a lot to parse, hopefully it all makes sense (and I hope I didn't miss anything!). Let me know if there are any follow-up questions.

**Comment by kuasha420** [comment]
@felixarntz Thank you for the detailed explanation and actions to move forward. I think I have done the steps as you've told and everything seems to be working. I have two more follow up questions:

>  In case of $cap being one of VIEW_DASHBOARD or VIEW_POSTS_INSIGHTS we need to call the above check_view_dashboard_capability( $user_id ), as both of these caps should basically rely on the same.

Should this change only happen when the `dashboardSharing` is enabled? Or can we apply this regardless? I think we should only run this when `dashboardSharing` is enabled right? 

> check_view_authenticated_dashboard_capability( $user_id ) method that simply checks if the user is authenticated and return array( AUTHENTICATE ) if yes or array( 'do_not_allow' ) if not.

This works for all cases except when the setup is incomplete. I have added `if ( $this->is_user_authenticated( $user_id ) && $this->authentication->is_setup_completed() ) { ... }` to account for those, is that the right place and/or the right thing to do? 

Cheers.

File: includes/Core/Permissions/Permissions.php [hunk 2]
```diff
@@ -163,31 +166,29 @@ public function __construct( Context $context, Authentication $authentication, M
...
 				$this->meta_to_base,
 				array(
-					// Allow users that can generally view the shared dashboard to read shared module data.
-					self::READ_SHARED_MODULE_DATA       => self::VIEW_SHARED_DASHBOARD,
+					// Allow users that can generally view dashboard to read shared module data.
+					self::READ_SHARED_MODULE_DATA       => self::VIEW_DASHBOARD,
 					// Admins who can manage options for SK can generally manage module sharing options.
 					self::MANAGE_MODULE_SHARING_OPTIONS => self::MANAGE_OPTIONS,
 					self::DELEGATE_MODULE_SHARING_MANAGEMENT => self::MANAGE_OPTIONS,
+					self::VIEW_SHARED_DASHBOARD         => self::VIEW_DASHBOARD,
```
Comment (felixarntz) [inline includes/Core/Permissions/Permissions.php:191]: Just wondering: Don't these arrows need to align for coding standards?

File: includes/Core/Permissions/Permissions.php [hunk 3]
```diff
@@ -399,6 +414,46 @@ private function check_dashboard_sharing_capability( $cap, $user_id, $args ) {
...
+	 * the authenticated or shared dashboard.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $user_id User ID of the user the capability is checked for.
+	 * @return array Array with a 'do_not_allow' element if checks fail, empty array if checks pass.
+	 */
+	private function check_view_dashboard_capability( $user_id ) {
+		if ( in_array( 'do_not_allow', $this->check_view_authenticated_dashboard_capability( $user_id ), true ) ) {
+			return $this->check_view_shared_dashboard_capability( $user_id );
```
Comment (felixarntz) [inline includes/Core/Permissions/Permissions.php:451]: Technically, we only need to call `check_view_shared_dashboard_capability` here if `dashboardSharing` is enabled, since otherwise there is no shared dashboard. Without the feature flag, we should technically return the result of `check_view_authenticated_dashboard_capability` here.

Basically, without the `dashboardSharing` flag, the `check_view_dashboard_capability` method should return exactly the same as `check_view_authenticated_dashboard_capability`.

This isn't really critical, since both `VIEW_DASHBOARD` and `VIEW_POSTS_INSIGHTS` are already covered by the more general condition in `map_meta_cap`, but probably a bit cleaner to make this method work for the current (without feature flag) requirements as well, just in case.

File: includes/Core/Permissions/Permissions.php [hunk 4]
```diff
@@ -428,6 +483,23 @@ private function check_view_shared_dashboard_capability( $user_id ) {
...
+	 *
+	 * Allows access to the VIEW_AUTHENTICATED_DASHBOARD capability if the user is authenticated.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $user_id User ID of the user the capability is checked for.
+	 * @return array Array with a 'do_not_allow' element if checks fail, otherise returns AUTHENTICATE capability.
+	 */
+	private function check_view_authenticated_dashboard_capability( $user_id ) {
+		if ( $this->is_user_authenticated( $user_id ) && $this->authentication->is_setup_completed() ) {
```
Comment (felixarntz) [inline includes/Core/Permissions/Permissions.php:497]: Good catch on adding the check for setup completed here. 👍 

The other cases are covered by that generic condition in our `map_meta_cap` callback, as almost all capabilities in `$base_to_core` or `$meta_to_core` will not be granted unless the setup is complete.

File: includes/Core/Permissions/Permissions.php [hunk 5]
```diff
@@ -360,6 +360,21 @@ private function map_meta_capabilities( array $caps, $cap, $user_id, $args ) {
...
+
+		// Handle VIEW_AUTHENTICATED_DASHBOARD capability.
+		if ( self::VIEW_AUTHENTICATED_DASHBOARD === $cap ) {
+			$caps = array_merge( $caps, $this->check_view_authenticated_dashboard_capability( $user_id ) );
+		}
+
+		// Special Handling of VIEW_DASHBOARD and VIEW_POSTS_INSIGHTS capabilities when the dashboardSharing feature flag is enabled.
+		if ( Feature_Flags::enabled( 'dashboardSharing' ) && in_array( $cap, array( self::VIEW_DASHBOARD, self::VIEW_POSTS_INSIGHTS ), true ) ) {
+			$caps = array_merge( $caps, $this->check_view_dashboard_capability( $user_id ) );
+		}
```
Comment (felixarntz) [inline includes/Core/Permissions/Permissions.php:376]: We can remove the feature flag check here and instead have it just around the `check_view_shared_dashboard_capability` call within this method, since only that is the part that should be conditional to the `dashboardSharing` feature flag.

However, your current approach works too, since both `VIEW_DASHBOARD` and `VIEW_POSTS_INSIGHTS` are already covered by the above clause starting in line 342. So either works for me.

**Review by felixarntz — COMMENTED** [review]
@kuasha420 This mostly looks good now!
> > In case of $cap being one of VIEW_DASHBOARD or VIEW_POSTS_INSIGHTS we need to call the above check_view_dashboard_capability( $user_id ), as both of these caps should basically rely on the same.
>
> Should this change only happen when the `dashboardSharing` is enabled? Or can we apply this regardless? I think we should only run this when `dashboardSharing` is enabled right?

It depends. Technically either works, but I think it would be cleaner to have this apply regardless, _however_ we must instead wrap the call to `check_view_shared_dashboard_capability` in a flag check, because that only ever applies with the feature flag. I left more details in the below comments.

> This works for all cases except when the setup is incomplete. I have added `if ( $this->is_user_authenticated( $user_id ) && $this->authentication->is_setup_completed() ) { ... }` to account for those, is that the right place and/or the right thing to do?

Great catch, yeah that looks good.

File: includes/Core/Permissions/Permissions.php (hunk 2 above)
Comment (felixarntz) [reply includes/Core/Permissions/Permissions.php:191]: Weird. Anyway, not a big deal.

Comment (aaemnnosttv) [reply includes/Core/Permissions/Permissions.php:191]: Yes, but for some reason this is where PHPCS thinks they should be 🤷‍♂️

**Review by felixarntz — APPROVED** [review]
@kuasha420 Awesome!

**Comment by felixarntz** [comment]
@aaemnnosttv Can you take another look since you previously had requested changes?

File: includes/Core/Admin/Screens.php [hunk 6]
```diff
@@ -264,7 +264,7 @@ private function no_access_redirect_dashboard_to_splash() {
 		}
 
 		// Redirect to splash screen if user is allowed to authenticate.
```
Comment (aaemnnosttv) [inline includes/Core/Admin/Screens.php:266]: This comment is a bit stale now as the condition isn't specific to authentication anymore but also probably redundant as the code is clear enough on its own IMO 🙂 

```suggestion
```

File: includes/Core/Permissions/Permissions.php [hunk 7]
```diff
@@ -360,6 +360,21 @@ private function map_meta_capabilities( array $caps, $cap, $user_id, $args ) {
...
+
+		// Handle VIEW_AUTHENTICATED_DASHBOARD capability.
+		if ( self::VIEW_AUTHENTICATED_DASHBOARD === $cap ) {
+			$caps = array_merge( $caps, $this->check_view_authenticated_dashboard_capability( $user_id ) );
+		}
+
+		// Special Handling of VIEW_DASHBOARD and VIEW_POSTS_INSIGHTS capabilities when the dashboardSharing feature flag is enabled.
+		if ( in_array( $cap, array( self::VIEW_DASHBOARD, self::VIEW_POSTS_INSIGHTS ), true ) ) {
+			$caps = array_merge( $caps, $this->check_view_dashboard_capability( $user_id ) );
+		}
```
Comment (aaemnnosttv) [inline includes/Core/Permissions/Permissions.php:376]: How about using a `switch` here as well, similar to how `check_dashboard_sharing_capability`? The comments seem less necessary this way too.

<details>
<summary>diff</summary>

```diff
diff --git a/includes/Core/Permissions/Permissions.php b/includes/Core/Permissions/Permissions.php
index 8429863df7..b5504403e9 100644
--- a/includes/Core/Permissions/Permissions.php
+++ b/includes/Core/Permissions/Permissions.php
@@ -360,19 +360,18 @@ final class Permissions {
 			$caps = array_merge( $caps, $this->check_dashboard_sharing_capability( $cap, $user_id, $args ) );
 		}
 
-		// Handle VIEW_SPLASH capability.
-		if ( self::VIEW_SPLASH === $cap ) {
-			$caps = array_merge( $caps, $this->check_view_splash_capability( $user_id ) );
-		}
-
-		// Handle VIEW_AUTHENTICATED_DASHBOARD capability.
-		if ( self::VIEW_AUTHENTICATED_DASHBOARD === $cap ) {
-			$caps = array_merge( $caps, $this->check_view_authenticated_dashboard_capability( $user_id ) );
-		}
-
-		// Special Handling of VIEW_DASHBOARD and VIEW_POSTS_INSIGHTS capabilities when the dashboardSharing feature flag is enabled.
-		if ( in_array( $cap, array( self::VIEW_DASHBOARD, self::VIEW_POSTS_INSIGHTS ), true ) ) {
-			$caps = array_merge( $caps, $this->check_view_dashboard_capability( $user_id ) );
+		switch ( $cap ) {
+			case self::VIEW_SPLASH:
+				$caps = array_merge( $caps, $this->check_view_splash_capability( $user_id ) );
+				break;
+			case self::VIEW_AUTHENTICATED_DASHBOARD:
+				$caps = array_merge( $caps, $this->check_view_authenticated_dashboard_capability( $user_id ) );
+				break;
+			// Intentional fallthrough.
+			case self::VIEW_DASHBOARD:
+			case self::VIEW_POSTS_INSIGHTS:
+				$caps = array_merge( $caps, $this->check_view_dashboard_capability( $user_id ) );
+				break;
 		}
 
 		return $caps;
```

</details>

File: includes/Core/Permissions/Permissions.php [hunk 8]
```diff
@@ -399,6 +414,48 @@ private function check_dashboard_sharing_capability( $cap, $user_id, $args ) {
...
+	 * @return array Array with a 'do_not_allow' element if checks fail, empty array if checks pass.
+	 */
+	private function check_view_splash_capability( $user_id ) {
+		if ( Feature_Flags::enabled( 'dashboardSharing' ) ) {
+			if ( in_array( 'do_not_allow', $this->check_view_shared_dashboard_capability( $user_id ), true ) ) {
+				return array( self::AUTHENTICATE );
+			}
+			return array();
+		}
+		return array( self::AUTHENTICATE );
```
Comment (aaemnnosttv) [inline includes/Core/Permissions/Permissions.php:435]: Nit: This could be simplified a little by making the guard specific to the feature not enabled case which would swap the placement of the returns making for overall less indentation.

**Review by aaemnnosttv — APPROVED** [review]
Nice work @kuasha420 ! I left a few comments but mostly small polish nitpicks. My comments from before have all been addressed though 👍
